/**
   File : ref.proto
   Package : edg.ref

   This is where we define the basic structures we use throughout EDG for
   referencing different elements in a designs or libraries.

   We enforce certain structural properties by having a series of nested
   'steps' that determine the next step in a path reference.
 */

syntax = "proto3";
package edg.ref;

import "common.proto";
import "name.proto";

/** These are reserved terms that we'll end up using in various places.
    I'd rather have these in the block/link/bridges where they're going
    to exist, but that's not possible without polymorphism protibuf doesn't have */
enum Reserved {
  UNDEFINED = 0;

  /** Directions we could move */
  CONNECTED_LINK = 1;

  /** reserved parameters */
  IS_CONNECTED = 40;  // true implies CONNECTED_LINK resolves; not available on PortArray
//  EXISTS = 41;
  LENGTH = 42;  // available on PortArray and LinkArray
  ALLOCATE = 43;  // available on PortArray
  NAME = 44;
}

/** This is an actual step you take within the local
    environment.

    The directions encode the type of thing we are referencing,
    but to the user all of these look just like local variables */
message LocalStep {
  oneof step {
    Reserved reserved_param = 1;
	
    /**
     A local name is what something is called in the context of its parent,
     whether that parent is a namespace (as in the library) or some other
     element (as in a design or heirarchy block).

     localNames should have the following properties:

     - First char is a lower case letter
     - All other chars must be letters, numbers, '-', '<', '>'
     - lowerCamelCase is preffered, don't use any symbols in the name
       if possible.

    These are style guidelines, literally any string will work. */
    string name = 3;
  }
}

/** This is a path from a local context to some other local context.
    To be used as a reference. */
message LocalPath {
  repeated LocalStep steps = 1;

  edg.common.Metadata meta = 255;
}

/** This is a path to an element within a library from the root of
    a library. To be used as a way to reference such elements. */
message LibraryPath {
  edg.name.LibraryName start = 1;
  repeated edg.name.Namespace steps = 2;
  LocalStep target = 3;

  edg.common.Metadata meta = 255;
}